darts/models/forecasting/tft_model.py

```diff
@@ -0,0 +1,915 @@
+        encoder_lengths = torch.tensor(
+            [self.input_chunk_length] * x_cont_past.shape[dim_samples], 
+            dtype=past_target.dtype,
+            device=past_target.device
+        )
+        decoder_lengths = torch.tensor(
+            [self.output_chunk_length] * x_cont_future.shape[dim_samples], 
+            dtype=past_target.dtype,
+            device=past_target.device
+        )
+
+        timesteps = self.input_chunk_length + self.output_chunk_length
+        encoder_length = self.input_chunk_length
+        input_vectors_past = {
+            name: x_cont_past[..., idx].unsqueeze(-1)
+            for idx, name in enumerate(self.encoder_variables)
+        }
+        input_vectors_future = {
+            name: x_cont_future[..., idx].unsqueeze(-1)
+            for idx, name in enumerate(self.decoder_variables)
+        }
+
+        # Embedding and variable selection
+        if static_covariates is not None:
+            # TODO: impelement static covariates
+            # # static embeddings will be constant over entire batch
+            # static_embedding = {name: input_vectors[name][:, 0] for name in self.static_variables}
+            # static_embedding, static_covariate_var = self.static_covariates_vsn(static_embedding)
+            raise NotImplementedError('Static covariates have yet to be defined')
+        else:
+            static_embedding = torch.zeros((past_target.shape[0], self.hidden_size),
+                                           dtype=past_target.dtype,
+                                           device=past_target.device)
+
+            # this is only to interpret the output
+            static_covariate_var = torch.zeros((past_target.shape[0], 0),
+                                               dtype=past_target.dtype,
+                                               device=past_target.device)
+
+        if future_covariates is None and static_covariates is None:
+            raise NotImplementedError('make zero tensor if future covariates is None')
+
+        static_context_expanded = self.expand_static_context(
+            context=self.static_context_grn(static_embedding),
+            timesteps=timesteps
+        )
+
+        embeddings_varying_encoder = {name: input_vectors_past[name] for name in self.encoder_variables}
+        embeddings_varying_encoder, encoder_sparse_weights = self.encoder_vsn(
+            x=embeddings_varying_encoder,
+            context=static_context_expanded[:, :encoder_length],
+        )
+
+        embeddings_varying_decoder = {name: input_vectors_future[name] for name in self.decoder_variables}
+        embeddings_varying_decoder, decoder_sparse_weights = self.decoder_vsn(
+            x=embeddings_varying_decoder,
+            context=static_context_expanded[:, encoder_length:],
+        )
+
+        # LSTM
+        # calculate initial state
+        input_hidden = self.static_context_initial_hidden_lstm(static_embedding).expand(
```

Comment:
I guess this initial GRN pass on the LSTM initial state might make sense, but i believe it's not in the TFT paper?

Reply:
The GRN comes from the static covariate encoders (you can see them in the bottom right of the image below in the other comment). It's not obvious from the image but when you read through the text:

From TFT paper:
![image](https://user-images.githubusercontent.com/84317528/140476600-363c874b-9b0c-4e42-bdc8-97b6f78954c9.png)
and for the encoder LSTM:
![image](https://user-images.githubusercontent.com/84317528/140476680-11c8ed27-a61c-4a10-8ae2-0e552293a4b3.png)